lib/ohai/plugins/linux/network.rb

```diff
@@ -406,17 +412,25 @@ def parse_ip_addr_inet6_line(cint, iface, line)
             ]
           end.first
 
-          unless route.nil? or route.empty?
+          if route && !route.empty?
             if family[:name] == "inet"
               ipaddress route[:src]
-              macaddress iface[route[:dev]][:addresses].select{|k,v| v["family"]=="lladdr"}.first.first unless iface[route[:dev]][:flags].include? "NOARP"
+              m = get_mac_for_interface(iface, route[:dev])
+              Ohai::Log.debug("Overwriting macaddress #{macaddress} with #{m} from interface #{route[:dev]}") if macaddress
+              macaddress m
             else
               ip6address route[:src]
+              if macaddress
+                Ohai::Log.debug("Not setting macaddress from ipv6 interface #{route[:dev]} because macaddress is already set")
+              else
+                macaddress get_mac_for_interface(iface, route[:dev])
+              end
             end
+          else
+            Ohai::Log.debug("Unable to determine ideal default route, not setting ipaddress/ip6address/macaddress. Should have considered #{default_route[:dev]}?")
```

Comment:
Can you identify any real scenario where we would get to this code path? If not, we're solving a problem that doesn't exist. I don't think we can get here, but I don't feel certain enough about that to fail to make someone report that to me.

It is absolutely okay to not pick a macaddress in some network configurations.

You need to be able to leave macaddress set to nil if the interface you're picking ipaddress from has no MAC. For a contrived case, if you run on a linux system with no routes and no external interfaces, you'll get nothing from `linux/network` because we have no routes, and fall back to the `network` plugin where you will get `ipaddress => 127.0.0.1` and `macaddress => nil` because the localhost interface has no need for layer 2 addressing. You wouldn't hit the code path we're discussing though because you don't have any routes.

If you go way back to [https://tickets.opscode.com/browse/OHAI-362](OHAI-362) we were breaking people who were using tun interfaces because we assumed macaddress would never be nil. Part of the fix of OHAI-362 is visible in the conditional check for the NOARP flag that still exists:

`macaddress iface[route[:dev]][:addresses].select{|k,v| v["family"]=="lladdr"}.first.first unless iface[route[:dev]][:flags].include? "NOARP"`

A point-to-point link isn't going to have a mac address because there's no addressing required. There's only one other host on the link, so anything we put on the wire is destined for that host and it expects everything it reads on the wire came from us.